lib/_http_outgoing.js

```diff
@@ -109,35 +109,29 @@ OutgoingMessage.prototype.setTimeout = function(msecs, callback) {
 // any messages, before ever calling this.  In that case, just skip
 // it, since something else is destroying this connection anyway.
 OutgoingMessage.prototype.destroy = function(error) {
-  if (this.socket)
+  if (this.socket) {
+    // If the socket is corked from a write, uncork it before destroying it.
+    if (this.socket._writableState.corked)
+      this.socket.uncork();
```

Comment:
> The behavior of destroy() is something we should define and clarify throughout core.

Not so much `destroy()`, it's just the `socket.write()` [API is weird](https://nodejs.org/api/net.html#net_socket_write_data_encoding_callback). It returns a boolean about whether or not it works like a sync or async function. It doesn't necessarily have to be *fixed*, it just has to be worked around because [`test/parallel/test-http-abort-client.js`](https://github.com/nodejs/node/blob/6de9d13f83e297f5347ed0a8dbfdcfc967da5a0a/test/parallel/test-http-abort-client.js#L8-L10) and userland doesn't always use it perfectly. I'll add a better comment.

> It is entirely possible that _writableState.corked is greater that one.

Ah, I think you're right. I'll fix that.